File: lib/src/zone.dart
import 'dart:math' as math;

import 'models.dart';
import 'i18n.dart';

/// How the zone is bounded.
enum RangeMode {
  /// A plain circle, in kilometres, as the crow flies. Exact.
  radius,

  /// A travel time, converted to a circle with average door-to-door speeds.
  travel,
}

/// How the user would get to the appointment.
enum TravelMode { walk, bike, transit, car }

extension TravelModeInfo on TravelMode {
  String get label => switch (this) {
        TravelMode.walk => tr.travelWalk,
        TravelMode.bike => tr.travelBike,
        TravelMode.transit => tr.travelTransit,
        TravelMode.car => tr.travelCar,
      };

  /// "en bus", "a pied": reads naturally after a duration.
  String get phrase => switch (this) {
        TravelMode.walk => tr.phraseWalk,
        TravelMode.bike => tr.phraseBike,
        TravelMode.transit => tr.phraseTransit,
        TravelMode.car => tr.phraseCar,
      };

  /// Average door-to-door speed in km/h along the actual route, in a French
  /// city. Deliberately conservative: better to include a practitioner who
  /// turns out 5 minutes further than to miss one.
  double get speedKmh => switch (this) {
        TravelMode.walk => 4.5,
        TravelMode.bike => 14,
        TravelMode.transit => 15,
        TravelMode.car => 28,
      };

  /// Fixed minutes lost regardless of distance: walking to the stop and
  /// waiting for the bus, parking the car, locking the bike.
  double get overheadMinutes => switch (this) {
        TravelMode.walk => 0,
        TravelMode.bike => 2,
        TravelMode.transit => 10,
        TravelMode.car => 5,
      };
}

/// Streets are not straight lines. Road distance is typically 1.2-1.4 times
/// the straight-line distance in a French city.
const double kDetourFactor = 1.3;

/// Estimated minutes to cover a straight-line distance.
double estimateMinutes(double straightKm, TravelMode mode) =>
    mode.overheadMinutes + straightKm * kDetourFactor / mode.speedKmh * 60;

/// Straight-line radius reachable within [minutes], the inverse of
/// [estimateMinutes]. Never below 300 m, so a very short budget still means
/// "around here" rather than "nowhere".
double radiusForMinutes(int minutes, TravelMode mode) {
  final usable = minutes - mode.overheadMinutes;
  final km = usable * mode.speedKmh / 60 / kDetourFactor;
  return math.max(0.3, km);
}

/// Great-circle distance in kilometres.
double haversineKm(double lat1, double lng1, double lat2, double lng2) {
  const r = 6371.0;
  double rad(double d) => d * math.pi / 180;
  final dLat = rad(lat2 - lat1);
  final dLng = rad(lng2 - lng1);
  final a = math.pow(math.sin(dLat / 2), 2) +
      math.cos(rad(lat1)) * math.cos(rad(lat2)) * math.pow(math.sin(dLng / 2), 2);
  return 2 * r * math.asin(math.sqrt(a));
}

/// The point [km] away from ([lat], [lng]) in the direction [bearingDeg].
({double lat, double lng}) offsetPoint(
  double lat,
  double lng,
  double km,
  double bearingDeg,
) {
  const r = 6371.0;
  final d = km / r;
  final b = bearingDeg * math.pi / 180;
  final p1 = lat * math.pi / 180;
  final l1 = lng * math.pi / 180;
  final p2 = math.asin(
    math.sin(p1) * math.cos(d) + math.cos(p1) * math.sin(d) * math.cos(b),
  );
  final l2 = l1 +
      math.atan2(
        math.sin(b) * math.sin(d) * math.cos(p1),
        math.cos(d) - math.sin(p1) * math.sin(p2),
      );
  return (lat: p2 * 180 / math.pi, lng: l2 * 180 / math.pi);
}

/// A town searched on behalf of the zone, before it is resolved to a
/// Doctolib place.
class Commune {
  const Commune({
    required this.name,
    required this.code,
    required this.lat,
    required this.lng,
    required this.population,
  });

  final String name;
  final String code;
  final double lat;
  final double lng;
  final int population;

  Map<String, dynamic> toJson() => {
        'name': name,
        'code': code,
        'lat': lat,
        'lng': lng,
        'population': population,
      };

  factory Commune.fromJson(Map<String, dynamic> j) => Commune(
        name: j['name'] as String,
        code: j['code'] as String? ?? '',
        lat: (j['lat'] as num).toDouble(),
        lng: (j['lng'] as num).toDouble(),
        population: (j['population'] as num?)?.toInt() ?? 0,
      );
}

/// An optional area around the user. Without one, an alert covers the whole
/// city it was created for; with one, it covers exactly the practitioners
/// inside the area, even across town borders.
class Zone {
  Zone({
    required this.lat,
    required this.lng,
    required this.label,
    this.mode = RangeMode.radius,
    this.radiusKm = 3,
    this.minutes = 30,
    this.travel = TravelMode.transit,
    this.places = const [],
    this.communes = const [],
  });

  /// The starting point: home, current position, or a point on the map.
  double lat;
  double lng;
  String label;

  RangeMode mode;
  double radiusKm;
  int minutes;
  TravelMode travel;

  /// Doctolib places searched for this zone, one per town it touches.
  List<PlaceRef> places;

  /// The towns themselves, kept for display.
  List<Commune> communes;

  /// The straight-line radius actually enforced.
  double get effectiveRadiusKm =>
      mode == RangeMode.radius ? radiusKm : radiusForMinutes(minutes, travel);

  bool contains(double lat, double lng) =>
      haversineKm(this.lat, this.lng, lat, lng) <= effectiveRadiusKm;

  double distanceKm(double lat, double lng) =>
      haversineKm(this.lat, this.lng, lat, lng);

  /// "3 km autour de" / "30 min en transports depuis".
  String get summary {
    if (mode == RangeMode.radius) {
      return tr.zoneAround(formatKm(radiusKm), label);
    }
    return tr.zoneFromTravel(minutes, travel.phrase, label);
  }

  /// Short form for chips: "3 km", "30 min en transports".
  String get shortLabel => mode == RangeMode.radius
      ? formatKm(radiusKm)
      : tr.zoneShortTravel(minutes, travel.phrase);

  /// How long the trip to a point at [km] would take, for display.
  String travelHint(double km) {
    final t = mode == RangeMode.travel ? travel : TravelMode.transit;
    final m = estimateMinutes(km, t).round();
    return tr.tripMinutes(m, t.phrase);
  }

  Map<String, dynamic> toJson() => {
        'lat': lat,
        'lng': lng,
        'label': label,
        'mode': mode.name,
        'radiusKm': radiusKm,
        'minutes': minutes,
        'travel': travel.name,
        'places': places.map((p) => p.toJson()).toList(),
        'communes': communes.map((c) => c.toJson()).toList(),
      };

  factory Zone.fromJson(Map<String, dynamic> j) => Zone(
        lat: (j['lat'] as num).toDouble(),
        lng: (j['lng'] as num).toDouble(),
        label: j['label'] as String? ?? '',
        mode: RangeMode.values.firstWhere(
          (m) => m.name == j['mode'],
          orElse: () => RangeMode.radius,
        ),
        radiusKm: (j['radiusKm'] as num?)?.toDouble() ?? 3,
        minutes: (j['minutes'] as num?)?.toInt() ?? 30,
        travel: TravelMode.values.firstWhere(
          (t) => t.name == j['travel'],
          orElse: () => TravelMode.transit,
        ),
        places: (j['places'] as List? ?? [])
            .map((e) => PlaceRef.fromJson(e as Map<String, dynamic>))
            .toList(),
        communes: (j['communes'] as List? ?? [])
            .map((e) => Commune.fromJson(e as Map<String, dynamic>))
            .toList(),
      );
}

String formatKm(double km) => I18n.km(km);

File: drlib_web/src/doctolib/zone.ts
/** Port of lib/src/zone.dart — the geometry half, without the UI labels. */

import type { PlaceRef } from './models';

export type RangeMode = 'radius' | 'travel';
export type TravelMode = 'walk' | 'bike' | 'transit' | 'car';

/**
 * Average door-to-door speed in km/h along the actual route, in a French city.
 * Deliberately conservative: better to include a practitioner who turns out
 * five minutes further than to miss one.
 */
const SPEED_KMH: Record<TravelMode, number> = {
  walk: 4.5,
  bike: 14,
  transit: 15,
  car: 28,
};

/**
 * Fixed minutes lost regardless of distance: walking to the stop and waiting
 * for the bus, parking the car, locking the bike.
 */
const OVERHEAD_MIN: Record<TravelMode, number> = {
  walk: 0,
  bike: 2,
  transit: 10,
  car: 5,
};

/**
 * Streets are not straight lines. Road distance is typically 1.2-1.4 times the
 * straight-line distance in a French city.
 */
export const DETOUR_FACTOR = 1.3;

/** Estimated minutes to cover a straight-line distance. */
export function estimateMinutes(straightKm: number, mode: TravelMode): number {
  return OVERHEAD_MIN[mode] + (straightKm * DETOUR_FACTOR) / SPEED_KMH[mode] * 60;
}

/**
 * Straight-line radius reachable within `minutes`, the inverse of
 * [estimateMinutes]. Never below 300 m, so a very short budget still means
 * "around here" rather than "nowhere".
 */
export function radiusForMinutes(minutes: number, mode: TravelMode): number {
  const usable = minutes - OVERHEAD_MIN[mode];
  const km = (usable * SPEED_KMH[mode]) / 60 / DETOUR_FACTOR;
  return Math.max(0.3, km);
}

/** Great-circle distance in kilometres. */
export function haversineKm(lat1: number, lng1: number, lat2: number, lng2: number): number {
  const R = 6371.0;
  const rad = (d: number) => (d * Math.PI) / 180;
  const dLat = rad(lat2 - lat1);
  const dLng = rad(lng2 - lng1);
  const a =
    Math.sin(dLat / 2) ** 2 +
    Math.cos(rad(lat1)) * Math.cos(rad(lat2)) * Math.sin(dLng / 2) ** 2;
  return 2 * R * Math.asin(Math.sqrt(a));
}

/** The point `km` away from (lat, lng) in the direction `bearingDeg`. */
export function offsetPoint(
  lat: number,
  lng: number,
  km: number,
  bearingDeg: number,
): { lat: number; lng: number } {
  const R = 6371.0;
  const d = km / R;
  const b = (bearingDeg * Math.PI) / 180;
  const p1 = (lat * Math.PI) / 180;
  const l1 = (lng * Math.PI) / 180;
  const p2 = Math.asin(Math.sin(p1) * Math.cos(d) + Math.cos(p1) * Math.sin(d) * Math.cos(b));
  const l2 =
    l1 +
    Math.atan2(
      Math.sin(b) * Math.sin(d) * Math.cos(p1),
      Math.cos(d) - Math.sin(p1) * Math.sin(p2),
    );
  return { lat: (p2 * 180) / Math.PI, lng: (l2 * 180) / Math.PI };
}

/** A town searched on behalf of the zone, before it is resolved to a place. */
export interface Commune {
  name: string;
  code: string;
  lat: number;
  lng: number;
  population: number;
}

/**
 * An optional area around the user. Without one, an alert covers the whole city
 * it was created for; with one, it covers exactly the practitioners inside the
 * area, even across town borders.
 */
export interface Zone {
  lat: number;
  lng: number;
  label: string;
  mode: RangeMode;
  radiusKm: number;
  minutes: number;
  travel: TravelMode;
  /** Doctolib places searched for this zone, one per town it touches. */
  places: PlaceRef[];
  /** The towns themselves, kept for display. */
  communes: Commune[];
}

/** The straight-line radius actually enforced. */
export function effectiveRadiusKm(z: Zone): number {
  return z.mode === 'radius' ? z.radiusKm : radiusForMinutes(z.minutes, z.travel);
}

export function zoneContains(z: Zone, lat: number, lng: number): boolean {
  return haversineKm(z.lat, z.lng, lat, lng) <= effectiveRadiusKm(z);
}

export function zoneDistanceKm(z: Zone, lat: number, lng: number): number {
  return haversineKm(z.lat, z.lng, lat, lng);
}

export function normaliseZone(raw: any): Zone | null {
  if (!raw || typeof raw !== 'object') return null;
  const lat = Number(raw.lat);
  const lng = Number(raw.lng);
  if (!Number.isFinite(lat) || !Number.isFinite(lng)) return null;
  const travel: TravelMode = ['walk', 'bike', 'transit', 'car'].includes(raw.travel)
    ? raw.travel
    : 'transit';
  return {
    lat,
    lng,
    label: String(raw.label ?? ''),
    mode: raw.mode === 'travel' ? 'travel' : 'radius',
    radiusKm: Number.isFinite(Number(raw.radiusKm)) ? Math.min(50, Math.max(0.3, Number(raw.radiusKm))) : 3,
    minutes: Number.isFinite(Number(raw.minutes)) ? Math.min(180, Math.max(5, Number(raw.minutes))) : 30,
    travel,
    places: Array.isArray(raw.places) ? raw.places : [],
    communes: Array.isArray(raw.communes) ? raw.communes : [],
  };
}
